src/Compilers/CSharp/Portable/Errors/DiagnosticInfoWithSymbols.cs

```diff
@@ -20,5 +21,17 @@ internal DiagnosticInfoWithSymbols(bool isWarningAsError, ErrorCode errorCode, o
         {
             this.Symbols = symbols;
         }
+
+        public override bool Equals(object obj)
+        {
+            return obj is DiagnosticInfoWithSymbols diws && 
+                   this.Symbols.Equals(diws.Symbols) && 
+                   base.Equals(obj);
+        }
```

Comment:
> Are we handling new DiagnosticInfo().Equals(new DiagnosticInfoWithSymbols()) as well?
  
We are "handling this" in the DiagnosticInfo.Equals method by simply returning `false` if `GetType` returns differently for either object being compared.

http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/DiagnosticInfo.cs,429

So, to answer your question, `a.Equals(b) == b.Equals(a)` for all `a` and `b` that are subclasses of DiagnosticInfo